### ISSUE TITLE
Cloud SDK not working together with CAP after upgrade to version 4.0.0

### ISSUE THREAD

**Issue by adamandre79** [SAP/cloud-sdk#1099]
Hello colleagues,

our project dependencies:

| Dependency | Version |
|:------------------|---------------------------------------|
| CAP Java SDK      | com.sap.cds: cds-services-bom; cds-adapter-odata-v2; cds-starter-spring-boot; cds-starter-cloudfoundry: 1.28.0;  |
| OData version     | v2                               |
| Node.js version   | ^16                            |
| @sap/cds            | 6.2.1                           |
| @sap/cds-mtx    | 2.6.0                           |
| Cloud SDK          | 4.0.0                           |

we have upgraded our cloud SDK to version 4.0.0 and have created an own async thread decorator like it was described here:

https://github.com/SAP/cloud-sdk/blob/main/docs-java/features/multi-tenancy/thread-context.mdx

After that our Spring Boot tests all works but after running it in our pipeline we got a error for the same process we have running in our Spring Boot tests:

```
org.springframework.transaction.CannotCreateTransactionException: Could not open JDBC Connection for transaction; nested exception is com.sap.cds.services.utils.ErrorStatusException: Ensure authenticated user context is provided and contains tenant information
```

We have no idea how we can avoid this error or what causes the problem.
As far as we can see the error occur in an async thread where we try to write something on the database within a new transaction.

Could you please help us to solve the problem?

What we also saw is that the is a decorator for SAP CAP context: CdsThreadContextDecorator and we think this is also registered. At least in our Spring Boot tests we see this decorator registered.

Here is how it looks like:

```
@Configuration
public class RequestContextExecutorConfiguration {

    public static final String REQUEST_CONTEXT_EXECUTOR = "requestContextExecutor";
    private static final Logger logger = LoggerFactory.getLogger(RequestContextExecutorConfiguration.class);

    @Bean(name = REQUEST_CONTEXT_EXECUTOR)
    @RequestScope
    public Executor requestContextExecutor() {
        DefaultThreadContextDecoratorChain.registerDefaultDecorator(new Decorator());
        return DefaultThreadContextExecutorService.of(Executors.newCachedThreadPool());
    }

    static class Decorator implements ThreadContextDecorator {

        @Override
        public int getPriority() {
            return DefaultPriorities.CUSTOM_DECORATOR;
        }

        @Nonnull
        @Override
        public <T> Callable<T> decorateCallable(@Nonnull Callable<T> callable) {
            var contextMap = MDC.getCopyOfContextMap();
            var securityContext = SecurityContextHolder.getContext();

            return () -> {
                setMdcContext(contextMap);
                logger.info("set MDC in new thread: {}", contextMap);
                SecurityContextHolder.setContext(securityContext);
                logger.info("set security context in new thread");
                return callable.call();
            };
        }

        private void setMdcContext(Map<String, String> contextMap) {
            if (contextMap != null) {
                MDC.setContextMap(contextMap);
            }
        }
    }

}
```

We opened a issue at SAP CAP with an equal description but they answered:

-----------------------------------------------------------------------------------------------

Hi @Andre Adam,

:::tip The SAP Cloud SDK already comes with a default decorator that passes on the SecurityContext of com.sap.cloud.security:java-api. If the [CAP integration](https://github.com/SAP/cloud-sdk/blob/main/docs-java/guides/cap-sdk-integration.mdx) is used, also the RequestContext is passed on (requires CDS version 1.28 or higher). :::

passing the security context comes already with the SDK. And passing the CAP context is already handled by the CdsThreadContextDecorator which comes with the CAP cloud-sdk integration. So normally, the only thing you need in your custom decorator is setting the MDC context.

Can you try to reduce your custom decorator to only that (setting MDC context) and see if you're still faced with the issue?

-----------------------------------------------------------------------------------------------

We tried it but got this error:

```
 method ... raised exception org.springframework.security.access.AccessDeniedException
org.springframework.security.access.AccessDeniedException: Access forbidden: not authenticated
	at com.sap.cloud.security.xsuaa.token.SpringSecurityContext.getToken(SpringSecurityContext.java:38)
```

After that answer from us we got this answer:

-----------------------------------------------------------------------------------------------

I think you should report this issue at the [cloud sdk support channel](https://github.com/SAP/cloud-sdk/issues/new/choose).

If the security context is not propagated properly without a custom decorator, as claimed in their documentation, that seems to be a bug on their side.

Also, either the executor bean that you've defined in your RequestContextExecutorConfiguration is not utilized when you use the async API or your custom decorator is not registered correctly. In any case they should support you with this as it's not part of the CAP Java API.

On CAP Java side the integration part we provide is the CdsThreadContextDecorator - our own custom decorator that is used by their aysnc API. The only difference I see is that we don't register it with Spring means, but with the help of our ApplicationLifecycleService.

------------------------------------------------------------------------------------------------

So could you please help us to solve this issue?

Regards
André

**Comment by Johannes-Schneider** [comment]
Hi @adamandre79,

thanks for reaching out to us.

Would you mind attaching your Maven dependency tree (simply run `mvn dependency:tree` in your project's root folder) to this issue?
I'd like to check whether all your dependency versions are as we would expect them to be.

Thanks and best regards,
Johannes

**Comment by MatKuhr** [comment]
In addition to Johannes comment: I see you are setting `@RequestScope` for your executor. Is this really what you want? Typically one can use just one executor that is being used for all requests. Using a new one per request can come with problems, e.g. you might have to protect your app against too many incoming requests to avoid your machine running out of threads / out of heap space. So Just asking to make sure this is indeed intentional.

**Comment by adamandre79** [comment]
Hello,

here is the dependency tree:
[tree.txt](https://github.com/SAP/cloud-sdk/files/9917328/tree.txt)

I can try to remove the request scope. I think this was needed with cloud SDK 3.x.x where we needed to set request context per http request.

Regards
André

**Comment by newtork** [comment]
Hi André,

In your dependency tree I've noticed...
```
 +- com.sap.cloud.sdk.cloudplatform:cloudplatform-connectivity:jar:4.0.0:compile
 ...
 +- com.sap.cds:cds-integration-cloud-sdk:jar:1.28.0:compile
 |  \- com.sap.cloud.sdk.frameworks:resilience4j:jar:3.75.0:compile
```

Can you please add the dependency directly or as dependency management entry to use...?
```
com.sap.cloud.sdk.frameworks:resilience4j:4.0.0
```

---

Furthermore I think there is a miscommunication.

* The following usage is for the Spring `SecurityContext` with namespace `org.springframework.security.core.context.SecurityContextHolder` - right?
  ```
  var securityContext = SecurityContextHolder.getContext();
  ...
  SecurityContextHolder.setContext(securityContext);
  ```
* Supposedly the CAP colleagues assumed you meant the `SecurityContext` with namespace `com.sap.cloud.security.token`. The Cloud SDK indeed already takes care of getting/setting this data.

Kind regards
Alexander

**Comment by adamandre79** [comment]
Hi Alexander,

I will check the dependencies und correct them.
One question to the security context. We are using this security library:

```
        <dependency>
            <groupId>com.sap.cloud.security.xsuaa</groupId>
            <artifactId>xsuaa-spring-boot-starter</artifactId>
            <version>2.13.0</version>
        </dependency>
```

Do I understand correctly, that you are supporting not this but `com.sap.cloud.security:java-api` out of the box and because we are not using `com.sap.cloud.security:java-api` we need to set it manually?

Regards
André

**Comment by adamandre79** [comment]
Hi Alexander,

from the SAP CAP colleagues we now get the information, that we shall try to run our custom decorator before there `CdsThreadContextDecorator`.  They wrote:

When the CdsThreadContextDecorator runs its decorateCallable, by creating the RequestContextRunnerInstance it will read the authentication information from the Spring's SecurityContext as well, and make it available in the CAP context.

I will also try this.

But does this mean, that your `SecurityContextThreadContextDecorator` also needs to run before the `CdsThreadContextDecorator`? Otherwise I would assume the security context is not set.

Regards
André

**Comment by newtork** [comment]
Hi André,

I looked into your PR `Update sap.cloud.sdk.version to v4 (major) #5041` and I found the change confusing since there are lots of side commits and mixed-in works.

My suggestion would be:

* (Separate PR) (Cloud SDK v3) **First** handle migration of (deprecated) Service SDK API to Cloud SDK API.
  e.g. `com.sap.cloud.sdk.odatav2.connectivity.ODataQueryResult` to `com.sap.cloud.sdk.datamodel.odata.client.request.ODataRequestResultGeneric`

* (Separate PR) Fix unrelated code changes
  * DestinationHandler.java
  * ProcedureRunResultsBlockingRequestController.java
  * DefaultProcedureRunInternalCloser.java
  * AbstractProcedureRunProcessor.java
  * ProcedureRunExecutionInitializer.java
  * RequestMessageTextsAccessor.java
  * DefaultProcedureRunDbUpdater.java
  * (maybe test filess)

* (This PR) (Cloud SDK v4) Leverage `sdk-bom` for dependency management to avoid dependency conflicts during update.
  * Check whether `RequestContextExecutorConfiguration` should really hold state for `CdsRuntime`. This is request scoped information, right? How does it fit into a `@Configuration` with presumably application scope.

Kind regards
Alexander

**Comment by adamandre79** [comment]
Hello Alexander,

I was not aware, that `ODataRequestResultGeneric` is also available with cloud SDK 3.x.x. Knowing this I can separate the pull requests like you suggested.
I will try this, thanks for the info.

For the sdk-bom: is there a information page for this?

For the `CdsRuntime`: I would like to remove this, but if we only set the security context and MDC we always get an error from the SAP CAP runtime, that the tenant information is missing and no transaction can be created.
I would hope that the `CdsRuntime` has a proxy mode and can so by used in a singleton. If not, I have no idea how to set the context for the CAP runtime to find the tenant information.

Regards
André

**Comment by Johannes-Schneider** [comment]
Hi @adamandre79,

> For the sdk-bom: is there a information page for this?

Sure. You can find more details [in our documentation](https://sap.github.io/cloud-sdk/docs/java/guides/manage-dependencies)

> For the CdsRuntime: I would like to remove this, but if we only set the security context and MDC we always get an error from the SAP CAP runtime, that the tenant information is missing and no transaction can be created.

That is surprising because the CDS context should be propagated automatically when using the `cds-integration-cloud-sdk` dependency.
Could you attach the exact exception (as well as the stacktrace) for the error you are observing?

Thanks and best regards,
Johannes


PS: I took a brief look into your PR to update to Cloud SDK 4.0 and left a comment with regards to the `testutil-core` dependency.

**Comment by adamandre79** [comment]
Hello @Johannes-Schneider ,

with the sdk-bom it looks good now. At the moment I get no error. 
So somehow the versions where mixed up and so some problems occur.

One problem we had with CAP that an extended entity which we create with help of the CAP sidecar was not found.
After we found a possible reason we introduced some extra logic for this and hope that is solved with this.

Our productive pipeline in running at the moment. If there no errors come up I would say the problem is solved.

If will update and perhaps close this issue if the pipeline is green.

Regards
André
